### PR TITLE
Add function restrictions

### LINKED ISSUE #1567: Require that vertex shaders return a position?
MSL requires that if a vertex shader does not return void, then it must at least return the position builtin. We should consider whether WGSL should impose the same requirement, and also whether it should be valid for a vertex shader to return `void`.

I'd like to propose that WGSL should require vertex shader return types to include `[[builtin(position)]]` if they are not `void`. I do not have a strong opinion on whether or not we should allow a return type of `void`.

### PR DESCRIPTION
**Pull request by alan-baker**
Fixes #1567, #1471, #1457, #1139

* No recursion
* No calling entry points
* fragment shader must return position
* aliasing restrictions
* pointer restrictions

The allowance handle pointer parameters is somewhat debatable, but I've included it since we don't have a clear alternative at the moment.

### CODE REVIEW

File: wgsl/index.bs [hunk 1]
```diff
@@ -5085,15 +5085,32 @@ Note: The current function will not resume execution if the called function or
...
-* Recursion is not permitted. (No cycle in the call graph.)
-* Function call parameters
- * Match type and number
- * Restrictions on pointers
- * Aliasing (?)
+## Restrictions ## {#function-restriction}
 
+* Recursion is not permitted.
+    That is, there must be no cycles in the call graph.
+* A [=fragment=] shader must return the `position` [=built-in output variable|built-in variable=].
```
Comment (jrprice) [inline wgsl/index.bs:5092]: fragment -> vertex

**Review by kvark — APPROVED** [review]

File: wgsl/index.bs [hunk 2]
```diff
@@ -5085,15 +5085,32 @@ Note: The current function will not resume execution if the called function or
...
- * Restrictions on pointers
- * Aliasing (?)
+## Restrictions ## {#function-restriction}
 
+* Recursion is not permitted.
+    That is, there must be no cycles in the call graph.
+* A [=vertex=] shader must return the `position` [=built-in output variable|built-in variable=].
+    See [[#builtin-variables]].
+* An entry point must never be the target of a [=function call=].
+* Functions must not return a [=pointer type=].
```
Comment (dneto0) [inline wgsl/index.bs:5095]: Can't return a handle either.
I think this is better said as:   If a function has a return type, it must be a plain type.

File: wgsl/index.bs [hunk 3]
```diff
@@ -5085,15 +5085,32 @@ Note: The current function will not resume execution if the called function or
...
+    * [=storage classes/function=]
+    * [=storage classes/private=]
+    * [=storage classes/handle=]
+    * [=storage classes/workgroup=]
+* Each argument of a function call of pointer type must be one of:
+    * An [[#address-of-expr|address-of expression]] of a
+        [[#var-identifier-expr|variable identifier expression]]
+    * A function parameter
+* [SHORTNAME] assumes no aliasing is present in the program. As such,
+    a function parameter of pointer type must not read or write to any
```
Comment (dneto0) [inline wgsl/index.bs:5107]: The subject is a little confusing; the pointer doesn't do the writing.
Suggest

> A function parameter of pointer type must not be used to read or write to any memory locations of its originating variable that are also written via:

**Review by dneto0 — CHANGES_REQUESTED** [review]
Looks great.

I've got two nits.

File: wgsl/index.bs [hunk 4]
```diff
@@ -5085,15 +5085,33 @@ Note: The current function will not resume execution if the called function or
...
+    the following storage classes:
+    * [=storage classes/function=]
+    * [=storage classes/private=]
+    * [=storage classes/handle=]
+    * [=storage classes/workgroup=]
+* Each argument of a function call of pointer type must be one of:
+    * An [[#address-of-expr|address-of expression]] of a
+        [[#var-identifier-expr|variable identifier expression]]
+    * A function parameter
+* [SHORTNAME] assumes no aliasing is present in the program. As such,
```
Comment (RobinMorisset) [inline wgsl/index.bs:5106]: Technically I think that we can have aliasing within a function, it is only across function boundaries that we cannot have aliasing, is that right?
e.g.
```
fn foo () {
    X = ...;
    const ptr: ptr<...> = &X;
    *ptr = ...;
}
```
is OK even though X is accessed in two ways. Or did I misunderstand that.

Comment (alan-baker) [reply wgsl/index.bs:5105]: Yes, that is ok. I'll change this to say:
> WGSL assumes no aliasing is present between any combination of variables and function parameters. As such, ...

Does that sounds clearer?

**Comment by jdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/WGSL-2021-04-27-Minutes#add-function-restrictions-1674">WGSL meeting minutes 2021-04-27</a></summary>



*   AB: The contentious part is whether you can pass a handle by pointer.  I left that in.   Otherwise, the restrictions reflect basic SPIR-V.
*   MM: At office hours, discussed that maybe if we allow texture pointers, it might break some of the analysis we want to do
*   DN: For texture filtering? (yes) How does that leak into WGSL?
*   MM: If possible to pass by ref, at the point that you compile, the pipeline layout tells you the inner type of the texture and whether it’s sampleable. If textures are named specifically, we can do this validation.
*   DM: Yes we need to know what textures are used by which samplers
*   MM: Can we add more types for this?
*   DM: We had that proposal earlier, and we didn’t want that at that time
*   DN: What’s proposal for passing tex/samplers to functions?
*   MM: By pointer would be great. If we want to keep the analysis, I think we wouldn’t be able to do this, rather you’d have to name it. I wish we could fix this. Statically determining tex/sampler at compiler time is valuable.
*   KN: I thought it wasn’t just analysis, but also some hardware needs to be able to determine this.
*   JG: Well, spirv isn’t really that flexible. Can determine static-use
*   KN: Could do e.g. unrolling to figure these out
*   MM: Let’s see if we can figure this out offline
*   DN: Can we land this without the contested bits?
*   


</details>

File: wgsl/index.bs (hunk 4 above)
Comment (dneto0) [reply wgsl/index.bs:5105]: I think that is better, and addresses @RobinMorisset's point.

I find the way explanation of the rule a bit confusing, so I suggest adding an Issue to revisit the expression of the rule.

**Review by dneto0 — APPROVED** [review]
Thanks!

**Pull request merged**